app/Modules/Availability/AvailabilityService.php

```diff
@@ -0,0 +1,177 @@
+        $avail->where(function ($query) use (&$from, &$to) {
+            // Where the start date is inside the booked time
+            $query->where(function ($query) use (&$from, &$to) {
+                $query->where('from', '>', $from)
+                    ->where('from', '<', $to);
+                // Or where the end date is inside the booked time
+            })->orWhere(function ($query) use (&$from, &$to) {
+                $query->where('to', '>', $from)
+                    ->where('to', '<', $to);
+                // Or where the times are the same
+            })->orWhere(['from' => $from, 'to' => $to]);
+        });
+
+        return $avail->doesntExist();
+    }
+
+    public function createAvailability(array $availabilityData): Availability
+    {
+        ['from' => $from, 'to' => $to, 'user_id' => $userId] = $availabilityData;
+
+        $availabilityUser = $this->user::findOrFail($userId);
+
+        try {
+            $passesTimeChecks = $this->validateAvailabilityTimes($from, $to, $availabilityUser);
+        } catch (Exception $e)  {
+            if (get_class($e) == 'AvailabilityInPastException') {
+                throw new AvailabilityInPastException();
+            } else {
+                throw new AvailabilityMinimumTimeException();
+            }
+        }
+
+        if (!$passesTimeChecks) {
+            throw new OverlappingAvailabilityException();
+
+        }
+        return $availabilityUser->availability()->create([
+            'user_id' => $availabilityUser->id,
+            'from' => $from,
+            'to' => $to,
+        ]);
+    }
+
+    public function updateExistingAvailability(array $newData, bool $adminOverride = false): bool
+    {
+        ['from' => $from, 'to' => $to, 'user_id' => $userId] = $newData;
+
+        /** @var Availability $existingAvailability */
+        $existingAvailability = Availability::findOrFail($newData['id']);
+
+        $availabilityUser = $this->user::findOrFail($userId);
+
+        if (!$adminOverride && ($userId != $existingAvailability['user_id'])) {
+            throw new PrivilegeException('update this availability');
+        }
+
+        try {
+            $passesTimeChecks = $this->validateAvailabilityTimes($from, $to, $availabilityUser, $existingAvailability->getKey());
+        } catch (Exception $e)  {
+            if (get_class($e) == 'AvailabilityInPastException') {
+                throw new AvailabilityInPastException();
+            } else {
+                throw new AvailabilityMinimumTimeException($this->TIME_IN_MINS);
+            }
+        }
+
+        if (!$passesTimeChecks) {
+            throw new OverlappingAvailabilityException();
+        }
+
+        return $existingAvailability->update([
+            'user_id' => $availabilityUser->id,
+            'from' => $from,
+            'to' => $to,
+        ]);
+    }
+
+    public function removeAvailabilityRange(array $newData, bool $adminOverride = false): int
```

Comment:
Don't need an `$adminOverride` here. If we needed an override at a later date, would most likely be in its own method with its own logic.

Reply:
So this is only going to be for a user removing/adding/modifying their own availability?

Reply:
either way it doesn't really matter (but realistically yes) - if we wanted to run it for a different user we can pass in the user ID in the data array.

Reply:
Yeah but that's why i put the admin override in, because you can pass a user_id in that isn't your own and it will query against that user

`
       if (!$adminOverride && ($userId != $existingAvailability['user_id'])) {
            throw new PrivilegeException('update this availability');
        }
`